File: packages/sdk-csharp/Ubag.Sdk/IdempotencyKey.cs
using System.Security.Cryptography;

namespace Ubag.Sdk;

/// <summary>Generates ULID-style idempotency keys: 26 Crockford base32 characters.</summary>
public static class IdempotencyKey
{
    private const string CrockfordBase32 = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";

    public static string Generate()
    {
        var nowMillis = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        Span<char> buffer = stackalloc char[26];
        EncodeBase32(nowMillis, buffer[..10]);

        Span<byte> random = stackalloc byte[16];
        RandomNumberGenerator.Fill(random);
        for (var i = 0; i < 16; i++)
        {
            buffer[10 + i] = CrockfordBase32[random[i] & 31];
        }

        return new string(buffer);
    }

    private static void EncodeBase32(long value, Span<char> target)
    {
        var remaining = value < 0 ? 0 : value;
        for (var i = target.Length - 1; i >= 0; i--)
        {
            target[i] = CrockfordBase32[(int)(remaining % 32)];
            remaining /= 32;
        }
    }
}

File: packages/sdk-php/src/IdempotencyKey.php
<?php

declare(strict_types=1);

namespace Ubag\Sdk;

/** Generates ULID-style idempotency keys: 26 Crockford base32 characters. */
final class IdempotencyKey
{
    private const CROCKFORD_BASE32 = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';

    public static function generate(?int $nowMillis = null): string
    {
        $nowMillis ??= (int) (microtime(true) * 1000);
        $key = self::encodeBase32(max($nowMillis, 0), 10);
        for ($i = 0; $i < 16; $i++) {
            $key .= self::CROCKFORD_BASE32[random_int(0, 31)];
        }

        return $key;
    }

    private static function encodeBase32(int $value, int $length): string
    {
        $buffer = array_fill(0, $length, '0');
        $remaining = $value;
        for ($i = $length - 1; $i >= 0; $i--) {
            $buffer[$i] = self::CROCKFORD_BASE32[$remaining % 32];
            $remaining = intdiv($remaining, 32);
        }

        return implode('', $buffer);
    }
}
